### ISSUE TITLE
Assigning name to index does not propagate it into partition's indexes

### ISSUE THREAD

**Issue by gshimansky** [modin-project/modin#1618]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:

Ubuntu 20.04

- **Modin version** (`modin.__version__`):

0.7.3+118.gf2cc740

- **Python version**:

Python 3.8.2

- **Code we can use to reproduce**:

```
import modin.pandas as pd
import numpy as np

df = pd.DataFrame({
    "A": [11, 21, 31, 41],
    "B": [12, 22, 32, 42]
})
df2 = pd.DataFrame({"A": [1.0, 0.0], "B": [0.0, 1.0]})

print(df)
print(df.index)
df.index.name = "name"
print(df.index)
if hasattr(df, "_query_compiler"):
    print(df._query_compiler._modin_frame._partitions[0][0].data.index)
print(df2)
print(df2.index)
df3 = df2.head(2)
print(df3)
print(df3.index)
df4 = df.dot(df3.T)
print(df4)
print(df4.index)
```

### Describe the problem

Code above assigns name to dataframe index, but it is only assigned to dataframe field `index`, not to its partition's indexes. When `dot` operation is performed, function `_apply_full_axis` is called which creates result's index from dataframe partitions' indexes in this code
```python
        if new_index is None:
            new_index = self._frame_mgr_cls.get_indices(
                0, new_partitions, lambda df: df.index
            )
```
Since index name is not set in partitions, result index name is not set as well.

**Comment by gshimansky** [comment]
Workaround like `df.index = df.index` after setting index name seems to help.

**Comment by pyrito** [comment]
As of latest master, it seems like the index name is not propagated to the partition still. This seems like relatively low priority since it's not breaking anything that we know of.

**Comment by gshimansky** [comment]
The code in this bug is a real code from a data scientist's notebook which broke on Modin. True that it has an easy workaround, but to use it a data scientist has to guess why his code works on Pandas but doesn't work on Modin.

**Comment by mvashishtha** [comment]
This is a tricky bug. We have to somehow associate the `index` of every modin pandas API object with the modin object that it comes from. Every time we mutate such index, we have to propagate the update (perhaps in a deferred manner) to the partitions of the modin object. In pandas, `index` is always just a pointer to an index object, so pandas doesn't face this problem.

**Comment by pyrito** [comment]
@mvashishtha yeah the issue is that we can't track any mutations to the Index. I wonder if we could address this issue by adding our own `ModinIndex` to wrap the index and defer updates to an underlying pandas access or something like that? Might be a can of worms, but could open the door to more optimizations with the Index?
